### PR TITLE
fundamental: Add SHA1

### PR DESCRIPTION
**Pull request by anonymix007**
Imported from libxcrypt

Split out from #34158

### CODE REVIEW

File: LICENSES/README.md [hunk 1]
```diff
@@ -66,6 +66,9 @@ The following exceptions apply:
    - src/libsystemd/sd-journal/lookup3.h
  * the tools/chromiumos/gen_autosuspend_rules.py script is licensed under the
    **BSD-3-Clause** license.
+ * the following sources are under **Public Domain** (LicenseRef-alg-sha1-public-domain):
+   - src/fundamental/sha1-fundamental.c
+   - src/fundamental/sha1-fundamental.c
```
Comment (bluca) [inline LICENSES/README.md]: `.h`

**Comment by poettering** [comment]
btw, the gcc fix should be squashed into the sha1 thing before merge, we want "perfect" PRs, after all, that are fully bisectable.

**Comment by anonymix007** [comment]
Yeah, I haven't squashed it only because it'll be easier to replace it with something else

**Comment by anonymix007** [comment]
@bluca:
Regarding your suggestion about reproducing, it actually fails in a completely unrelated place:
```
In file included from ../../../../src/basic/alloc-util.h:276,
                 from ../../../../src/libsystemd-network/sd-dhcp6-client.c:13:
In function ‘memcpy_safe’,
    inlined from ‘client_append_oro’ at ../../../../src/libsystemd-network/sd-dhcp6-client.c:707:17,
    inlined from ‘dhcp6_client_send_message’ at ../../../../src/libsystemd-network/sd-dhcp6-client.c:815:13:
../../../../src/basic/memory-util.h:27:16: error: ‘memcpy’ writing 18446744073709551614 bytes into a region of size 1 [-Werror=stringop-overflow=]
   27 |         return memcpy(dst, src, n);
      |                ^~~~~~~~~~~~~~~~~~~
In function ‘malloc_multiply’,
    inlined from ‘client_append_oro’ at ../../../../src/libsystemd-network/sd-dhcp6-client.c:703:21,
    inlined from ‘dhcp6_client_send_message’ at ../../../../src/libsystemd-network/sd-dhcp6-client.c:815:13:
../../../../src/basic/alloc-util.h:119:16: note: destination object of size 1 allocated by ‘malloc’
  119 |         return malloc(size * need ?: 1);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

File: LICENSES/README.md [hunk 2]
```diff
@@ -66,6 +66,9 @@ The following exceptions apply:
    - src/libsystemd/sd-journal/lookup3.h
  * the tools/chromiumos/gen_autosuspend_rules.py script is licensed under the
    **BSD-3-Clause** license.
+ * the following sources are under **Public Domain** (LicenseRef-alg-sha1-public-domain):
```
Comment (bluca) [inline LICENSES/README.md:69]: please also add the license text to `LICENSES/alg-sha1-public-domain.txt`

**Comment by poettering** [comment]
did you tried the `#pragma` solution i suggested?

**Comment by anonymix007** [comment]
> did you tried the `#pragma` solution i suggested?

Yes, just pushed, but I actually can't compile it locally because of the error above

**Comment by anonymix007** [comment]
`#pragma` solution doesn't work in clang:
```
../src/fundamental/sha1-fundamental.c:259:17: error: unknown warning group '-Wstringop-overread', ignored [-Werror,-Wunknown-warning-option]
                DISABLE_WARNING_STRINGOP_OVERREAD;
                ^
../src/fundamental/macro-fundamental.h:37:9: note: expanded from macro 'DISABLE_WARNING_STRINGOP_OVERREAD'
        _Pragma("GCC diagnostic ignored \"-Wstringop-overread\"")
        ^
<scratch space>:88:25: note: expanded from here
 GCC diagnostic ignored "-Wstringop-overread"
                        ^
1 error generated.
```

File: src/test/test-sha1.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                        sha1_process_string(str + i, &ctx);
+                        sha1_finish_ctx(&ctx, result + j);
+
+                        hex_result = hexmem(result + j, SHA1_DIGEST_SIZE);
+                        ASSERT_STREQ(hex_result, expect);
+                }
+        }
+}
+
+
```
Comment (poettering) [inline src/test/test-sha1.c]: double empty line

**Comment by poettering** [comment]
> `#pragma` solution doesn't work in clang:
> 
> ```
> 
> ```

christ. maybe add -Wunknown-warning-option to basic_disabled_warnings in meson.build to systematically get rid of that problem.

File: src/fundamental/sha1-fundamental.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        z += (w ^ x ^ y) + blk(i) + 0x6ED9EBA1 + rol(v, 5); \
+        w = rol(w, 30);
+#define R3(v, w, x, y, z, i)                                              \
+        z += (((w | x) & y) | (w & x)) + blk(i) + 0x8F1BBCDC + rol(v, 5); \
+        w = rol(w, 30);
+#define R4(v, w, x, y, z, i)                                \
+        z += (w ^ x ^ y) + blk(i) + 0xCA62C1D6 + rol(v, 5); \
+        w = rol(w, 30);
+
+
```
Comment (yuwata) [inline src/fundamental/sha1-fundamental.c]: doubled empty line.

File: src/fundamental/sha1-fundamental.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        state[1] += b;
+        state[2] += c;
+        state[3] += d;
+        state[4] += e;
+
+        /* Wipe variables */
+        a = b = c = d = e = 0;
+}
+
+
```
Comment (yuwata) [inline src/fundamental/sha1-fundamental.c]: Also here.

File: src/fundamental/sha1-fundamental.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        /* SHA1 initialization constants */
+        ctx->state[0] = 0x67452301;
+        ctx->state[1] = 0xEFCDAB89;
+        ctx->state[2] = 0x98BADCFE;
+        ctx->state[3] = 0x10325476;
+        ctx->state[4] = 0xC3D2E1F0;
+        ctx->count[0] = ctx->count[1] = 0;
+}
+
+
```
Comment (yuwata) [inline src/fundamental/sha1-fundamental.c]: Also here.

File: src/fundamental/sha1-fundamental.c [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                for (; i + 63 < size; i += 64)
+                        sha1_do_transform(ctx->state, (const uint8_t *) buffer + i);
+                REENABLE_WARNING;
+                j = 0;
+        } else
+                i = 0;
+        memcpy(&ctx->buffer[j], (const uint8_t *) buffer + i, size - i);
+}
+
+
```
Comment (yuwata) [inline src/fundamental/sha1-fundamental.c]: Also here.

File: src/fundamental/sha1-fundamental.c [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+By anonymix007 <48598263+anonymix007@users.noreply.github.com>
+Still 100% public domain
+modified for use with systemd
+*/
+
+#if SD_BOOT
+#  include "efi-string.h"
+#else
+#  include <string.h>
+#endif
```
Comment (yuwata) [inline src/fundamental/sha1-fundamental.c]: This is not necessary. Same block is in memory-util-fundamental.h .

**Review by yuwata — COMMENTED** [review]
Several minor comments.

File: meson.build [hunk 9]
```diff
@@ -369,6 +369,7 @@ basic_disabled_warnings = [
         '-Wno-missing-field-initializers',
         '-Wno-unused-parameter',
         '-Wno-nonnull-compare',
+        '-Wunknown-warning-option',
```
Comment (yuwata) [inline meson.build]: Maybe, `-Wno-unknown-warning-option` ? That is, missing `no-` ??

File: src/fundamental/sha1-fundamental.c (hunk 8 above)
Comment (anonymix007) [reply src/fundamental/sha1-fundamental.c]: It's also in string-util-fundamental.h and sha256-fundamental.c

**Review by yuwata — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**